controllers/opentelemetrycollector_controller.go

```diff
@@ -0,0 +1,156 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/open-telemetry/opentelemetry-operator/api/v1alpha1"
+	"github.com/open-telemetry/opentelemetry-operator/internal/config"
+	"github.com/open-telemetry/opentelemetry-operator/pkg/collector/reconcile"
+)
+
+// OpenTelemetryCollectorReconciler reconciles a OpenTelemetryCollector object
+type OpenTelemetryCollectorReconciler struct {
+	client.Client
+	log    logr.Logger
+	scheme *runtime.Scheme
+	config config.Config
+	tasks  []Task
+}
+
+// Task represents a reconciliation task to be executed by the reconciler
+type Task struct {
+	Name        string
+	Do          func(context.Context, reconcile.Params) error
+	BailOnError bool
+}
+
+// Params is the set of options to build a new openTelemetryCollectorReconciler
+type Params struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+	Config config.Config
+	Tasks  []Task
+}
+
+// NewReconciler creates a new reconciler for OpenTelemetryCollector objects
+func NewReconciler(p Params) *OpenTelemetryCollectorReconciler {
+	if len(p.Tasks) == 0 {
```

Comment:
Looks like it is used only for testing - is there an alternative way to set the test tasks?

Reply:
Following `kubebuilder`'s bootstrap code, the new tests are, whenever possible, using the `_test` package, so that they aren't able to use package-internal fields. In this case, it means that we can't set the internal `tasks` field without exposing it as a public field. Are you concern that this is leaking implementation details to callers, or do you have another concern as well?

Reply:
No, was just related to exposing impl details. Wonder why they changed the convention of having tests in the same package - or is this because some of the tests are also integration?

Reply:
They didn't, I assume this is one of the places where kubebuilder did differently than operator-sdk. We are still figuring out the tests, I'm open to changing it back to white-box testing.

Reply:
Tracking issue: https://github.com/open-telemetry/opentelemetry-operator/issues/34